l5kit/l5kit/evaluation/csv_utils.py

```diff
@@ -0,0 +1,246 @@
+
+import numpy as np
+
+MAX_MODES = 3
+
+"""
+These utilities write and read csv with ground-truth and prediction data.
+Both share the first two field (timestamp and track_id) which are used to identify a single record.
+
+GT:
+Single mode future prediction with availabilities (either 1->available or 0->unavailable).
+Header fields have these meanings:
+timestamp, track_id, avail_time_0, avail_time_1, ..., coord_x_time_0, coord_y_time_0, ...
+
+PRED:
+Multi mode future prediction with confidence score (one value per mode up to MAX_MODES, sum to 1).
+Header fields have these meanings:
+timestamp, track_id, conf_mode_0, conf_mode_1, ..., coord_x_time_0_mode_0, coord_y_time_0_mode_0, ...,
+coord_x_time_0_mode_1, coord_y_time_0_mode_1, ...
+
+"""
+
+
+def _generate_coords_keys(future_len: int, mode_index: int = 0) -> List[str]:
+    """
+    Generate keys like coord_x00, coord_y00... that can be used to get or set value in CSV.
+    Two keys for each mode and future step.
+
+    Args:
+        future_len (int): how many prediction in the future data has
+        mode_index (int): what mode are we reading/writing
+
+    Returns:
+        List[str]: a list of keys
+    """
+    return list(
+        chain.from_iterable([[f"coord_x{mode_index}{i}", f"coord_y{mode_index}{i}"] for i in range(future_len)])
+    )
+
+
+def _generate_avails_keys(future_len: int) -> List[str]:
+    """
+    Generate availabilities keys (one per future step)
+
+    Args:
+        future_len (int): how many prediction in the future data has
+
+    Returns:
+        List[str]: a list of keys
+    """
+    return [f"avail_{i}" for i in range(future_len)]
+
+
+def _generate_confs_keys() -> List[str]:
+    """
+    Generate modes keys (one per mode)
+
+    Returns:
+        List[str]: a list of keys
+    """
+    return [f"conf_{i}" for i in range(MAX_MODES)]
+
+
+def write_gt_csv(
+    csv_path: str, timestamps: np.ndarray, track_ids: np.ndarray, coords: np.ndarray, avails: np.ndarray
+) -> None:
+    """
+    Encode the ground truth into a csv file
+
+    Args:
+        csv_path (str): path to the csv to write
+        timestamps (np.ndarray): (num_example,) frame timestamps
+        track_ids (np.ndarray): (num_example,) agent ids
+        coords (np.ndarray): (num_example x future_len x num_coords) meters displacements
+        avails (np.ndarray): (num_example x future_len) array with value 0 (discard in eval) or 1 (keep in eval)
+
+    Returns:
+
+    """
+    assert len(coords.shape) == 3
+    num_example, future_len, num_coords = coords.shape
+    assert num_coords == 2
+    assert timestamps.shape == track_ids.shape == (num_example,)
+    assert avails.shape == (num_example, future_len)
+
+    coords_keys = _generate_coords_keys(future_len)
+    avails_keys = _generate_avails_keys(future_len)
+
+    # create and write HEADER
+    # order is (timestamp,track_id,availabilities,coords)
+    fieldnames = ["timestamp", "track_id"]
+    fieldnames.extend(avails_keys)
```

Comment:
Minor: Do you prefer this over ["timestamp", "track_id"] + avails_keys .... ?

Reply:
not really, I'm fine with replacing it :)